### ISSUE TITLE
Characteristic doesn't receive value

### ISSUE THREAD

**Issue by Pythonaire** [vojtamolda/homebridge-bluetooth#21]
I use ESP32, got BLE working correctly and send periodically Notification Values.
 
On a mobile with BLE Scanner: The sensor is seen, periodically NOTIFY update values are received.
On the raspberry: Sensor is connected, Service and Characteristic is "connected".
In the Homekit (also tested in the EVE App): Sensor is seen, correct service and characteristic. I can start "identify" the sensor with positive response. I see the action "Identify" in the Raspi log too.
 But ... never get values or value updates. I would expect the same behavior, like the BLE Scanner do. I'm confused. please help.

**Comment by vojtamolda** [comment]
Hello @Pythonaire, do you have the characteristic UUIDs correct in `config.json`?

**Comment by Pythonaire** [comment]
Yes, all things seems to be ok. (I use the same UUID's as you).  If I made an identity check on the EVE app back to the sensor, it works. No complaining messages in the debug log. I see, the Homebridge itself set the "event value" from true to false after some seconds. I thinks because of nothing happened. I pull BLE notify values from the sensor every 5 seconds ... nothing. No error messages, no warnings about things went wrong from the plugin. Just the 'ping pong' of HAP request ... response ... finished ... complete of the Homebridge itself.

**Comment by Pythonaire** [comment]
using gatttool on the raspberry to check:
the gatttool identified the characteristic by handler 2a for my characteristic UUID, the PROPERTY_NOTIFY flag  2b. By "char-write-req 2b 100" (100 means: switch on Notify) I can receive value updates. I think, the noble or the plugin should do the same, to receive Notify value.

**Comment by Pythonaire** [comment]
Again:
refresh the Homebridge by deleting persist and accessories. 
Then: got value one time.
Later: homebridge Processing characteristic set: [{"aid":2,"iid":10,"ev":false}] and "Unregistering Characteristic "Current Relative Humidity" for events"

**Comment by Pythonaire** [comment]
Maybe a step further: 
it seems, noble is a bit unstable. Got some irritating situations, responses while testing. At least, I used PROPERTY_READ and PROPERTY_NOTIFY to indicate value changes on the sensor. But nothing happens. Then, I added PROPERTY_INDICATE as BLECharacteristic and got values each time tapping the home app.   Unfortunately, response value  is 0... that's not right.

**Comment by vojtamolda** [comment]
Cool, @Pythonaire. Thanks for the updates. It looks like you're progressing. Honestly I'm not sure how to help you other than to wish you to have fun and good luck!

**Comment by Pythonaire** [comment]
got Homebridge message, that "updateReachability" is no longer supported by apple Homekit (contained in accessory.js). Next inject log messages (this.log.info(this.prefix, "blah, blah"), to see the flow of received notification values ... no luck. No error, no GET messages.
The only progress I see in the log is **"Accessory [Homebridge] Got Characteristic "Current Relative Humidity" value: 0"**. There is the point, I can see , the plugin should received values?
I hoped to see "this.log.warn(this.prefix, "Read from bluetooth characteristic failed | " + error);" [line 79] in the BluetoothCharacteristic.prototype.get function, if I made something wrong.

**Comment by Pythonaire** [comment]
sorry @vojtamolda, that I'm rude...I'm not really familiar with this.
... thinking about the "BluetoothCharacteristic.prototype.notify" in characteristic.js. I read, noble needs to be pointed to a notify event with "subscribe", mentioned above in the permission function. This is made on the sensor side with an descriptor. The sensor act as an BLE Server and send notify periodically. To read descriptor values, I think the noble function is "readValue". The   BluetoothCharacteristic.prototype.notify seems to read from buffer, but who filled the buffer?

**Comment by Pythonaire** [comment]
ok, maybe for later reading.

Found it: ....
All board you describe, using BLECurie for BLE communication. I use Sparkfun ESP32 Thing boards, they are smaller, and battery powered, but using other BLE Libraries. Sending the data via "(char*)&value, sizeof(value)"  solved the problems.

**Comment by vojtamolda** [comment]
I'm glad you found it. Is it a bug related to #22 that should be fixed? Do you want to do a PR?

**Comment by Pythonaire** [comment]
Yes it's related to #22. I use BLE libraries [https://github.com/nkolban/ESP32_BLE_Arduino](url), for my ESP32 Chip to build up my own sensors and try to bind these to the homekit over BLE.
Unfortunately, I 'm not really familiar with JavaScript and reverse engineering make It not better ,many pitfalls for me... So I though, the ESP BLE Libraries are the problem ... it's seems to be not.
Now, I found that  accessory.js use this.homebridgeAccessory.updateReachability(true), that its no longer supported. To pull a noble.read action, characteristic.js starts "this.homebridgeCharacteristic.on('get', this.get.bind(this));". 

I'm set UpdateReachability to false, change "homebridgeCharacteristic.on" to "homebridgeCharacteristic.updateValue", change "callback(null, value)" in "BluetoothCharacteristic.prototype.get" to "return value" and now I got values seen in the log file and through the homekit on my mobile.

Other issues I have at this time : two charactertistics on the ESP, they are connected but I got data just from one. The notification works by chance, I don't now why. As I describe, I'm not familiar with C++ programming, its more try and error.

**Comment by Pythonaire** [comment]
![image](https://user-images.githubusercontent.com/46537110/53526688-36a1b580-3ae5-11e9-8a5b-45a2c0f8d9d6.png)
--> step into the "BluetoothCharacteristic.prototype.connect" and found "PERMS.READ" Characteristic --> should call "this.get.bind(this))", but never reached.

**Comment by Pythonaire** [comment]
> I'm glad you found it. Is it a bug related to #22 that should be fixed? Do you want to do a PR?

Thx, yes its related to #22. Sorry for my impatience. 
Now it seems to be solved for me. I tried different methods to hand over data to the node.js buffer function. (the easiest and I think most stable is "std:string"). BUT using "sizeof(data)" with std::string leads into node.js crashes. Since I set fixed data length (ex. float with 4, instead of sizeof(float value)), it works.
